### ISSUE TITLE
Support for kamelet repository

### ISSUE THREAD

**Issue by lburgazzoli** [apache/camel-k#1818]
The current behaviour of the camel-k operator is to looks for kamelets from k8s but to support different use cases, we should think about introducing a concept of `kamelet repository`. As example we can support:

1. to search kamelet on the local filesystem (@doru1004 FYI)
2. to search kamelet from an online kamelet catalog (github)

**Comment by heiko-braun** [comment]
Maybe https://www.apicur.io/registry/ ?

**Comment by oscerd** [comment]
Not sure the apicurio registry is the right fit for this.

**Comment by doru1004** [comment]
What would be the relationship between a local kamelet and its online version? Is the local kamelet a struct subset of the online kamelet ?

**Comment by lburgazzoli** [comment]
> What would be the relationship between a local kamelet and its online version? Is the local kamelet a strict subset of the online kamelet ?

there should not be any difference, but if you are off-line or access to a cluster, then you may need a way to resolves kamelet if it is in use in your integration wit `kamel local run`

**Comment by doru1004** [comment]
Sounds good and I agree that there shouldn't be any difference. I was hinting at Kubernetes as a dependency. Could a kamelet run locally without Kubernetes being required?

**Comment by lburgazzoli** [comment]
1. no, the format should be the same 
2. a kamelet alone can't run so it need to be put in a context: a kamelet binding or plain integration as any other endpoint so it should not make any difference if youa re on or off k8s (of course if the kamelet itself required k8s then it wont' work but that's expected)

**Comment by heiko-braun** [comment]
@oscerd Why not? Isn't it just a place to put stuff? The examples I saw are basically just yaml or json documents.

**Comment by oscerd** [comment]
My main concern about the registry is related to the fact that we may need to provide a different strategy for the kamelets. I mean, in a schema registry, when you send a kafka record and you want to use a schema, you'll have to set a strategy to provide the id for that schema. Apicurio provides some strategies out of the box, but we may need to define our own strategy for the kamelets and I don't know where this strategy could live. But It may be I'm thinking about the schema registry as something only kafka-related...

**Comment by zregvart** [comment]
I guess the question would be: what is a _kamelet repository_ and what functionality is expected from it. Is it a dumb store of YAML/JSON files, similar to a Maven repository or a mounted volume. Or is it something with more functionality, similar to a CMIS repository.

**Comment by lburgazzoli** [comment]
> I guess the question would be: what is a _kamelet repository_ and what functionality is expected from it. Is it a dumb store of YAML/JSON files, similar to a Maven repository or a mounted volume. Or is it something with more functionality, similar to a CMIS repository.

I don't think it matters at the moment, the main goal is to introduce the concept on camel-k side, like defining an minimalistic API the operator can leverage to search and retrieve a kamelet. There will be multiple implementation depending on the back-end.

**Comment by lburgazzoli** [comment]
> @oscerd Why not? Isn't it just a place to put stuff? The examples I saw are basically just yaml or json documents. Take a look here: [apicur.io/registry/docs/apicurio-registry/1.3.2.Final/assets-attachments/registry-rest-api.htm](https://www.apicur.io/registry/docs/apicurio-registry/1.3.2.Final/assets-attachments/registry-rest-api.htm) It goves you an idea of the data model.
> 
> At the moment the registry doesn't have any notion of Kamelets, but it could I guess. There are some interesting traits:
> 
>     * validation
> 
>     * rules that govern modifcations
> 
>     * versioning
> 
>     * metadata
> 
> 
> I think all of these could be interesting to a registry that manages kamelets.

I think it would technically be possible and as you mentioned it would offer some nice things like version and validation, however I wonder if it would be semantically correct as a kamelet defines both the signature (with a json schema and the schema for in/out data) as well as the implementation of the API itself.

FWIW, I think we have or at least we thought about having an option to use the registry to store the schemas associated to the kamelet.

**Comment by nicolaferraro** [comment]
I'd start with a Github interface for the operator/CLI and also create a new Kamelet repository on the apache org that can be contributed by the community, like `apache/camel-kamelets`.

At the moment, it would contain only the yaml definitions, so that the runtime can look them up automatically. But the same repo can later host also other artifacts that are linked to each kamelet, like UIs configurators and related meta services.

The same repository can be also the source of truth for a new section on the website where users can look for kamelets and configure them for their platform (e.g. to create a Knative source, or to download it as Camel-Kafka-connector).

Wdyt?
